### ISSUE TITLE
Low TPS will cause Baritone to become stuck inside of a glitched block occasionally

### ISSUE THREAD

**Issue by DashGit** [cabaletta/baritone#161]
It should be noted that this issue really only affects servers like 2B2T or a low-FPS singleplayer world. However, it is still an issue, even if it is an edge case.

Sometimes when mining with high-efficiency tools, Baritone will get stuck inside of a block that it thinks it mined, but the server disagrees. This causes an infinite loop of recalculations so that the user has to manually cancel pathing and either break themselves out or reconnect to the server or world. Since this is mainly an issue with Minecraft and not Baritone's pathfinding, I would suggest simply stopping path recalculations when Baritone has tried and failed to move itself forward around 3 or so times.

**Comment by leijurv** [comment]
If this is on the same build as reported in #160 https://github.com/cabaletta/baritone/commit/72eec135d02d09b773240c175edb3c8accab8654, there was a bug that caused you to get stuck under an overhang indefinitely. 

However I just recorded another hour and a half today of 2b2t baritone impact at tps ranging from 1 to 4 and never hit this. And here's another 44 minutes yesterday (https://youtu.be/wvM04oKX6Kg), never hit this specifically. It failed to break a block sometimes but it never got permanently stuck. Actually, it may be because I know impact has built in anti desync though, and vanilla baritone runs into this.

**Comment by DashGit** [comment]
> If this is on the same build as reported in #160 [72eec13](https://github.com/cabaletta/baritone/commit/72eec135d02d09b773240c175edb3c8accab8654), there was a bug that caused you to get stuck under an overhang indefinitely.
> 
> However I just recorded another hour and a half today of 2b2t baritone impact at tps ranging from 1 to 4 and never hit this. And here's another 44 minutes yesterday (https://youtu.be/wvM04oKX6Kg), never hit this specifically. It failed to break a block sometimes but it never got permanently stuck. Actually, it may be because I know impact has built in anti desync though, and vanilla baritone runs into this. Were you using baritone + impact or just baritone?
> 
> Thanks for the report!

I've been using Baritone + Impact ever since I started using Baritone. What you were describing where it failed to break a block was the only issue I had on a multiplayer server that sort of mirrored what I described in the original post, so it may be a singleplayer only thing. I say this because where it got permanently stuck only happened on a singleplayer world with 32 chunks set as the rendering distance so TPS was pretty low. 

I didn't know Impact had anti-desync though - I checked and I didn't have it enabled because I've only been using auto-eat and ESP to test out Baritone. I'll try to get this to happen again with anti-desync enabled.

**Comment by leijurv** [comment]
Yes, there are some other known issues with auto eat that'll need integration. (for example, I just watched it sprint jump over a gap, then start auto eating in the middle of the jump, and fall haha).

Interesting, I've never actually used it in singleplayer at 32 chunks. What's your CPU and what TPS did you get with that?

The list of what impact utility modules I leave on can be found in this vid https://www.youtube.com/watch?v=wvM04oKX6Kg

However I'd probably err on the side of less rather than more for testing and reproducing issues.

**Comment by ZeroMemes** [comment]
This should be fixed in 1.13, and as mentioned, Anti Desync also fixes it.

```java
@Inject(
        method = "onPlayerDestroyBlock",
        at = @At(
                value = "INVOKE",
                target = "net/minecraft/world/World.setBlockState(Lnet/minecraft/util/math/BlockPos;Lnet/minecraft/block/state/IBlockState;I)Z"
        ),
        cancellable = true
)
private void onPlayerDestroyBlock$setBlockState(BlockPos pos, CallbackInfoReturnable<Boolean> ci) {
    if (shouldPreventDesync...) {
        ci.setReturnValue(false);
    }
}
```

**Comment by ZeroMemes** [comment]
> This should be fixed in 1.13, and as mentioned, Anti Desync also fixes it.
> 
> ```java
> @Inject(
>         method = "onPlayerDestroyBlock",
>         at = @At(
>                 value = "INVOKE",
>                 target = "net/minecraft/world/World.setBlockState(Lnet/minecraft/util/math/BlockPos;Lnet/minecraft/block/state/IBlockState;I)Z"
>         ),
>         cancellable = true
> )
> private void onPlayerDestroyBlock$setBlockState(BlockPos pos, CallbackInfoReturnable<Boolean> ci) {
>     if (shouldPreventDesync...) {
>         ci.setReturnValue(false);
>     }
> }
> ```
